### PR TITLE
Fix buoyancy

### PR DESCRIPTION
**Pull request by Thomio-Watanabe**
- Fix the buoyancy behavior when the models are rotated.
- Read the buoyancy as a input parameter, when defined within the gazebo world model.
- Calculates the viscous Drag.

### CODE REVIEW

File: src/GazeboUnderwater.cpp [hunk 1]
```diff
@@ -28,74 +26,110 @@ namespace gazebo_underwater
...
+        // buoyancy must be the buoyancy when the model is completely submersed
+        buoyancy = getParameter<double>("buoyancy","N",0);
     }
 
     void GazeboUnderwater::updateBegin(common::UpdateInfo const& info)
     {
-        physics::LinkPtr link = model->GetLink( sdf->Get<std::string>("link_name") ); 
-        applyBuoyancy(link);
-        applyViscousFriction(link);
+        if(link)
```
Comment (doudou) [inline src/GazeboUnderwater.cpp]: In which cases could the link not be present ? Shouldn't you make an error and make sure gazebo quits in this case ?

Comment (Thomio-Watanabe) [inline src/GazeboUnderwater.cpp]: This plugin will not find the link if one misspell the link name, i.e. if the link name inside the plugin tag doesn't match the link model name. (I actually need to test two things: if model_name is defined and if model_name name a model loaded in gazebo - todo).
I don't know if I should quit the simulation, because we can add a feature to apply buoyancy to multiple links/models. So, if we are testing the simulation we can comment a link/model inside the world file and still run the simulation. What is your opinion ?

Comment (doudou) [inline src/GazeboUnderwater.cpp]: My opinion is that "silently failing" or just warning in this case is source of trouble. If you comment the link, then comment the buyoancy

Comment (Thomio-Watanabe) [inline src/GazeboUnderwater.cpp]: ok, I'll change the code

**Comment by Thomio-Watanabe** [comment]
- test model and link elements
- calculate viscous drag
- use gazebo exception

File: src/GazeboUnderwater.cpp [hunk 2]
```diff
@@ -28,74 +27,133 @@ namespace gazebo_underwater
...
 
-        size = getParameter<math::Vector3>("size","meters",
-                math::Vector3(1, 1, 1));
-        centerOfBuoyancy = getParameter<math::Vector3>("center_of_buoyancy","meters",
-                math::Vector3(0, 0, 0.1));
-        fluidVelocity = getParameter<math::Vector3>("fluid_velocity","m/s",
-                math::Vector3(0,0,0));
-        viscousDamping = getParameter<math::Vector3>("viscous_damping","dimensionless",
-                math::Vector3(20,30,50));
         waterLevel = getParameter<double>("water_level","meters", 2.0);
```
Comment (doudou) [inline src/GazeboUnderwater.cpp]: a water level of 0 would seem like a more intuitive default

File: src/GazeboUnderwater.cpp [hunk 3]
```diff
@@ -28,74 +27,133 @@ namespace gazebo_underwater
...
-        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1000);
+        fluidVelocity = getParameter<math::Vector3>("fluid_velocity","m/s",math::Vector3(0,0,0));
+        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1027);
+        dragCoefficient = getParameter<math::Vector3>("drag_coefficient","dimensionless",
+                math::Vector3(1,1,1));
+        // buoyancy must be the buoyancy when the model is completely submersed
+        buoyancy = getParameter<double>("buoyancy","N",0);
+        centerOfBuoyancy = getParameter<math::Vector3>("center_of_buoyancy","meters",
+                math::Vector3(0, 0, 0.15));
+        sideAreas = getParameter<math::Vector3>("side_areas","meter2",
```
Comment (doudou) [inline src/GazeboUnderwater.cpp]: Can't you use the link's bounding box to compute that ? (I think I mentioned it in the email, but you did not answer)

Comment (Thomio-Watanabe) [inline src/GazeboUnderwater.cpp]: I was speaking with Jan and I think he wants this way.
The same for the volume variable. (because the real robot may have empty spaces inside it)

File: src/GazeboUnderwater.cpp (hunk 2 above)
Comment (Thomio-Watanabe) [inline src/GazeboUnderwater.cpp]: water level set to 0

File: src/GazeboUnderwater.cpp (hunk 3 above)
Comment (quams) [inline src/GazeboUnderwater.cpp]: Am 01.04.2015 um 17:33 schrieb Thomio Watanabe notifications@github.com:

> In src/GazeboUnderwater.cpp:
> 
> > ```
> >      waterLevel = getParameter<double>("water_level","meters", 2.0);
> > ```
> > -        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1000);
> > -        fluidVelocity = getParametermath::Vector3("fluid_velocity","m/s",math::Vector3(0,0,0));
> > -        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1027);
> > -        dragCoefficient = getParametermath::Vector3("drag_coefficient","dimensionless",
> > -                math::Vector3(1,1,1));
> > -        // buoyancy must be the buoyancy when the model is completely submersed
> > -        buoyancy = getParameter<double>("buoyancy","N",0);
> > -        centerOfBuoyancy = getParametermath::Vector3("center_of_buoyancy","meters",
> > -                math::Vector3(0, 0, 0.15));
> > -        sideAreas = getParametermath::Vector3("side_areas","meter2",
> 
> I was speaking with Jan and I think he wants this way.
> The same for the volume variable. (because the real robot may have empty spaces inside it)

Using the bounding box for the calculation only makes sense, if the volume of this box actually represents the volume of water displaced. In most underwater systems that is not entirely the case, especially with a covered open frame like FlatFish (or HROV for that matter).

So I thought its better to force the user to provide displaced Volume and weight as a parameter.

blue skies
    Jan

## 

-- Jan Christian Albiez -- jan@ankerwin.de -- http://www.ankerwin.de/ --
-- segmentation violation in module reality.o ...
                         ... please shutdown your universe and reboot --

Comment (doudou) [inline src/GazeboUnderwater.cpp]: I was referring to the sideAreas parameter, which are geometric properties. I do agree with the volume/weight thing. My comment was really ambiguous ...

Comment (Thomio-Watanabe) [inline src/GazeboUnderwater.cpp]: Jan @quams , this is your call.

Comment (doudou) [inline src/GazeboUnderwater.cpp]: Ping ...

File: src/GazeboUnderwater.cpp [hunk 4]
```diff
@@ -20,82 +20,140 @@ namespace gazebo_underwater
...
-            link->GetWorldPose().rot.RotateVectorReverse(fluidVelocity);
-        math::Vector3 velocityDifference =
-            link->GetRelativeLinearVel() - fluidVelocityRelative;
-        link->AddRelativeForce(- velocityDifference * viscousDamping);
+        math::Vector3 cogPosition = link->GetWorldCoGPose().pos;
+        double distanceToSurface = waterLevel - cogPosition.z;
+        if(distanceToSurface > 0 )
+        {
+            math::Vector3 velocityDifference = link->GetWorldCoGLinearVel() - fluidVelocity;
+            math::Vector3 viscousDrag = - 0.5 * densityOfFluid * sideAreas * dragCoefficient *
```
Comment (doudou) [inline src/GazeboUnderwater.cpp:108]: Don't you need to project (ortho) the bounding box w.r.t the direction of the flow ?

**Comment by doudou** [comment]
@Thomio-Watanabe: what would be the difference between the value of sideAreas and what we would get from the bounding box ?

**Comment by Thomio-Watanabe** [comment]
When we use bounding box to calculate the side areas we suppose the model is a rectangular box. 
But, likely the real side areas are smaller than the bounding box areas.

File: src/GazeboUnderwater.cpp (hunk 4 above)
Comment (Thomio-Watanabe) [inline src/GazeboUnderwater.cpp:108]: What do you mean ?

**Comment by doudou** [comment]
> When we use bounding box to calculate the side areas we suppose the model is a rectangular box. 
> But, likely the real side areas are smaller than the bounding box areas.

Fair enough. However, I really believe that you should at least offer the option of _by default_ using the bounding box. That would make using the plugin that much nicer.

**Comment by Thomio-Watanabe** [comment]
Ok, I will work on this

**Comment by Thomio-Watanabe** [comment]
Use BoudingBox size to calculate sideAreas by default

File: src/GazeboUnderwater.cpp [hunk 5]
```diff
@@ -20,82 +20,153 @@ namespace gazebo_underwater
...
+        // If side_areas are not given in world file we use the bouding box dimensions to calculate it
+        if(sdf->HasElement("side_areas"))
+        {
+            sideAreas = getParameter<math::Vector3>("side_areas","meter2",
+                    math::Vector3(0.0,0.0,0.0));
+            if( sideAreas == math::Vector3(0.0,0.0,0.0) )
+                gzthrow("GazeboUnderwater: side_areas cannot be (0.0, 0.0, 0.0).");
+        }else{
+            math::Box linkBoudingBox = link->GetBoundingBox();
+            sideAreas = getParameter<math::Vector3>("side_areas","meter2",
```
Comment (doudou) [inline src/GazeboUnderwater.cpp]: Isn't that "get parameter and use bounding box by default" ? If yes, why is there the if (hasElement("side_areas")) ?

Comment (Thomio-Watanabe) [inline src/GazeboUnderwater.cpp]: sorry I've gone crazy...

Comment (Thomio-Watanabe) [inline src/GazeboUnderwater.cpp]: In this case should I rewrite history (rebase) or make another commit ?

Comment (doudou) [inline src/GazeboUnderwater.cpp]: I would personally amend (i.e. rewrite the history). You don't need to rebase: since it is the last commit, you can amend it (there's an option for that in git gui)

**Comment by Thomio-Watanabe** [comment]
Amending last commit
